Using provider instead of HttpServletRequest in ServletJaxRsContextFactoryProvider

ganeshs opened the pull request:
This is the fix for https://github.com/pac4j/dropwizard-pac4j/issues/43

victornoel commented:
@ganeshs FYI I won't be able to review this right now, but as soon as I can I will experiment with it.

Could you explain briefly why you had to change things for resteasy, what are the new dependencies for and if it means that resteasy users need a special setup for pac4j to work now?

Thanks for your help and patience :)

ganeshs commented:
The change is to inject a javax.inject.provider instead httpservletrequest. For the resteasy module, I have modified the tests to use cdi. This means, the resteasy users will have to use cdi for injecting the provider. I have added a test that would act as an example for cdi integration.

ganeshs commented:
@victornoel any update on this

victornoel commented on core/src/main/resources/META-INF/beans.xml:
```diff
@@ -0,0 +1,9 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Could you move this with the test application? To me it has nothing to do here, it is only needed for resteasy with cdi, no? Or explain why it is needed here please

victornoel replied:
Ok, I agree, thank you for the explanation :)
Could you just add some comment in that file explaining why it is here exactly and maybe saying it is mainly because of resteasy/cdi?

victornoel commented on resteasy/src/test/java/org/pac4j/jax/rs/rules/RestEasyUndertowServletRule.java:
```diff
@@ -66,7 +72,13 @@ protected void before() throws Throwable {
+        ResteasyDeployment deployment = new ResteasyDeployment();
+        deployment.setInjectorFactoryClass("org.jboss.resteasy.cdi.CdiInjectorFactory");
+        deployment.setApplication(new MyApp());
+        DeploymentInfo di = server.undertowDeployment(deployment).setContextPath("/").setDeploymentName("DI")
```
improve indentation (one call per line)

victornoel commented on resteasy/src/test/java/org/pac4j/jax/rs/rules/RestEasyUndertowServletRule.java:
```diff
@@ -66,7 +72,13 @@ protected void before() throws Throwable {
 
-        server.deploy(new MyApp());
+        ResteasyDeployment deployment = new ResteasyDeployment();
+        deployment.setInjectorFactoryClass("org.jboss.resteasy.cdi.CdiInjectorFactory");
```
prefer to use `CdiInjectorFactory.class.getName()` instead

victornoel commented on resteasy/pom.xml:
```diff
@@ -45,11 +46,29 @@
+    </dependency>
+    <dependency>
+      <groupId>javax</groupId>
+      <artifactId>javaee-api</artifactId>
```
what is this for?

victornoel commented on resteasy/src/test/java/org/pac4j/jax/rs/rules/RestEasyUndertowServletRule.java:
```diff
@@ -66,7 +72,13 @@ protected void before() throws Throwable {
+        deployment.setInjectorFactoryClass("org.jboss.resteasy.cdi.CdiInjectorFactory");
+        deployment.setApplication(new MyApp());
+        DeploymentInfo di = server.undertowDeployment(deployment).setContextPath("/").setDeploymentName("DI")
+			.setClassLoader(RestEasyUndertowServletRule.class.getClassLoader())
```
prefer to use `getClass().getClassLoader()`

victornoel commented on resteasy/src/test/java/org/pac4j/jax/rs/rules/RestEasyUndertowServletRule.java:
```diff
@@ -66,7 +72,13 @@ protected void before() throws Throwable {
+        deployment.setApplication(new MyApp());
+        DeploymentInfo di = server.undertowDeployment(deployment).setContextPath("/").setDeploymentName("DI")
+			.setClassLoader(RestEasyUndertowServletRule.class.getClassLoader())
+			.addListeners(Servlets.listener(org.jboss.weld.environment.servlet.Listener.class));
```
do not use fqdn if possible, it is better to have all imports in import section so we can quickly see the dependencies of the class.

victornoel commented on core/src/main/java/org/pac4j/jax/rs/servlet/features/ServletJaxRsContextFactoryProvider.java:
```diff
@@ -17,11 +17,11 @@
-    @Context
-    private HttpServletRequest request;
+    @Inject
+    private javax.inject.Provider<HttpServletRequest> requestProvider;
```
You can now use an import instead :)

ganeshs commented:
@victornoel Addressed the changes requested.

victornoel commented on README.md:
```diff
@@ -36,7 +36,7 @@ These filters can be directly registered by hand, or instead, the following feat
 
 - The `Pac4JValueFactoryProvider` enables injection of the security profile in resource method (for Apache Jersey <2.26, see [#30](https://github.com/pac4j/jax-rs-pac4j/issues/30))
-- The `ServletJaxRsContextFactoryProvider` provides session handling (and thus indirect clients support) by replacing the generic `JaxRsContextFactoryProvider` (for Servlet-based JAX-RS implementations, e.g., Jersey on Netty or Grizzly Servlet, Resteasy on Undertow).
+- The `ServletJaxRsContextFactoryProvider` provides session handling (and thus indirect clients support) by replacing the generic `JaxRsContextFactoryProvider` (for Servlet-based JAX-RS implementations, e.g., Jersey on Netty or Grizzly Servlet, Resteasy on Undertow). This provider returns a RequestScoped context and so should be added as a class `ServletJaxRsContextFactoryProvider.class` instead  ...
```
Sorry, but that's not the right place where to do that.
This sentence should be kept as it is and the examples below should be updated as well as a little line of explanation about resteasy-cdi next to the section related to resteasy.

ganeshs replied:
Done. Hope this works.

victornoel commented:
Thanks @ganeshs, great work!
